### PR TITLE
Refactor currency and change fishing tokens to currency

### PR DESCRIPTION
**Pull request by Glamurio**
[Secret] [QoL] [Code Quality] [Internal]
## About the PR 
This PR changes `/obj/item/spacecash` to `/obj/item/currency/spacecash`, and does the same for all other variants. 
Additionally, it changes `/obj/item/requisition_token/fishing` to `/obj/item/currency/fishing`.

## Why's this needed? 
Currency has a lot of old, copy pasted code that was in need of a refactor. This was spurned on by the fact that fishing tokens right now do not stack, and are handled much more like a currency than a token. Because I wanted to avoid copy-pasting currency code another time, I felt it smarter to make a currency parent type for all existing and future currencies.

## Changelog 

```changelog
(u)Glamurio (Ryou)
(+)Fishing tokens now stack.
```

### CODE REVIEW

**Comment by Glamurio** [comment]
One day I will learn not to have 10 merge commits. One day.

**Comment by Gannets** [comment]
Thanks for this! I'll try and find time to test it.

File: code/obj/item/space_cash.dm [hunk 1]
```diff
@@ -42,219 +46,194 @@
...
-			user.visible_message("<span class='notice'>[user] stacks some cash.</span>")
+			user.visible_message("<span class='notice'>[user] stacks some currency.</span>")
 			stack_item(I)
 		else
 			..(I, user)
 
 	attack_hand(mob/user)
 		if ((user.l_hand == src || user.r_hand == src) && user.equipped() != src)
-			var/amt = round(input("How much cash do you want to take from the stack?") as null|num)
+			var/amt = round(input("How much currency do you want to take from the stack?") as null|num)
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:71]: As above.

File: code/obj/item/space_cash.dm [hunk 2]
```diff
@@ -266,11 +245,11 @@
 	get_desc()
 		. += "This one is worth [amount >= 1000000 ? "ONE FUCKING GOD DAMN MILLION" : amount] spacebux."
 
-	proc/setup(var/atom/L, var/amt = 1 as num)
+	setup(var/atom/L, var/amt = 1 as num)
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:248]:
```suggestion
	setup(atom/L, amt = 1)
```
Unnecessary

File: code/obj/item/space_cash.dm [hunk 3]
```diff
@@ -266,11 +245,11 @@
...
 		set_loc(L)
 		set_amt(amt)
 
-	proc/set_amt(var/amt = 1 as num)
+	set_amt(var/amt = 1 as num)
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:252]:
```suggestion
	set_amt(amt = 1)
```
Unnecessary

File: _std/copy.dm [hunk 4]
```diff
@@ -23,7 +23,7 @@ proc/semi_deep_copy(orig, new_arg=null, list/environment=null, root=null, copy_f
 			istype(orig, /datum/packet_network))
 		return orig
 	if(copy_flags & COPY_SKIP_EXPLOITABLE && (
-			istype(orig, /obj/item/uplink) || istype(orig, /obj/item/spacebux) || istype(orig, /obj/item/chem_hint) || istype(orig, /obj/item/pixel_pass)))
+			istype(orig, /obj/item/uplink) || istype(orig,  /obj/item/currency/spacebux) || istype(orig, /obj/item/chem_hint) || istype(orig, /obj/item/pixel_pass)))
```
Comment (frawhst) [inline _std/copy.dm:26]: It looks like you accidentally added a space before all the instances of `/obj/item/currency/spacebux`? Should be fixable via regex though.

File: code/obj/item/space_cash.dm [hunk 5]
```diff
@@ -382,55 +340,11 @@ TYPEINFO(/obj/item/stamped_bullion)
...
-				L.storage.add_contents(src)
-			else
-				src.set_loc(L)
-		set_amt(amt)
-
-	proc/set_amt(var/amt = 1 as num)
-		var/default_amount = default_min_amount == default_max_amount ? default_min_amount : rand(default_min_amount, default_max_amount)
-		src.amount = max(amt,default_amount)
-		src.UpdateStackAppearance()
+	stack_type = /obj/item/currency/fakecash // so all FAKE cash types can stack with each other
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:347]: this should probably be the same as `spacecash` for additional trickery
```suggestion
	stack_type = /obj/item/currency/fakecash // so all FAKE cash types can stack with each other
	display_name = "cash"
```

File: code/obj/item/space_cash.dm [hunk 6]
```diff
@@ -449,91 +363,104 @@ TYPEINFO(/obj/item/stamped_bullion)
...
+		default_max_amount = 1000000
+
+
+/obj/item/currency/fishing
+	name = "1 research ticket"
+	real_name = "research ticket"
+	icon_state = "fish_common"
+	desc = "A Nanotrasen aquatic research ticket compatible with the Fishing Equipment Vendor."
+	stack_type = /obj/item/currency/fishing // so all fishing tokens stack
+	default_min_amount = 1
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:418]:
```suggestion
	default_min_amount = 1
	display_name = "research tickets"
```

File: code/obj/item/space_cash.dm [hunk 7]
```diff
@@ -42,219 +47,195 @@
...
-	throwforce = 1
-	throw_speed = 1
-	throw_range = 8
-	w_class = W_CLASS_TINY
-	burn_possible = 0
-	amount = 1
-	max_stack = 1000000
-	stack_type = /obj/item/spacebux
-
+	stack_type = /obj/item/currency/spacebux
```
Comment (frawhst) [inline code/obj/item/space_cash.dm:238]:
```suggestion
	stack_type = /obj/item/currency/spacebux
	display_name = "spacebux"
```

**Review by frawhst — APPROVED** [review]
should just set `display_name` on the other types, otherwise lgtm. Gonna testmerge for a bit in case anything wacky happens.

**Pull request merged**